### PR TITLE
HDF5 reader/writer

### PR DESCRIPTION
**Pull request by simonbyrne**
Implement #747.
Work in progress

- [x] Code follows the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) OR N/A.
- [x] Unit tests are included OR N/A.
- [x] Code is exercised in an integration test OR N/A.
- [x] Documentation has been added/updated OR N/A.

### CODE REVIEW

File: src/IO/writers.jl [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+
+"""
+function write_new(
+    writer::HDF5Writer,
+    mesh::Meshes.IntervalMesh,
+    name::AbstractString = "interval",
+)
+    group = create_group(writer.file, name)
+    write_attribute(group, "type", "IntervalMesh")
+    write_new(writer, mesh.domain, "$name/domain")
```
Comment (simonbyrne) [inline src/IO/writers.jl:68]:
```suggestion
    write_attribute(group, "domain", HDF5.name(write(writer, mesh.domain))
```

File: src/IO/writers.jl [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+function write_new(
+    writer::HDF5Writer,
+    mesh::Meshes.IntervalMesh,
+    name::AbstractString = defaultname(mesh),
+)
+    group = create_group(writer.file, name)
+    write_attribute(group, "type", "IntervalMesh")
+    write_new(writer, mesh.domain, "$name/domain")
+    write_attribute(group, "nelements", length(mesh.faces))
+    return nothing
```
Comment (simonbyrne) [inline src/IO/writers.jl:90]: return `name`?

File: src/IO/writers.jl [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+    writer::HDF5Writer,
+    mesh::Meshes.RectilinearMesh,
+    name::AbstractString = defaultname(mesh),
+)
+    group = create_group(writer.file, "meshes/$name")
+    write_attribute(group, "type", "RectilinearMesh")
+    write_attribute(group, "domain", typeof(mesh.domain))
+    write_attribute(group, "interval1", "interval1") # TODO
+    write_attribute(group, "inverval2", "interval2") # TODO
+    return group
```
Comment (simonbyrne) [inline src/IO/writers.jl:103]: return `name`

File: src/IO/readers.jl [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+            DataLayout = eval(Meta.parse(data_layout))
+            ElType = eval(
+                Meta.parse(replace(attrs(g)["value_type"], "ClimaCore." => "")),
+            )
+            values = DataLayout{ElType, Nij}(data)
+            spacename =
+                occursin("c", fieldname) ?
+                "CenterExtrudedFiniteDifferenceSpace" :
+                "FaceExtrudedFiniteDifferenceSpace"
+            cache[fieldname] = Fields.Field(values, cache[spacename])
```
Comment (simonbyrne) [inline src/IO/readers.jl:170]: We probably don't need to cache fields

File: src/IO/readers.jl [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+    return nothing
+end
+
+function read_fieldvector(reader::HDF5Reader)
+    read_domains(reader)
+    read_meshes(reader)
+    read_topologies(reader)
+    read_spaces(reader)
+    read_fields(reader)
+    return Fields.FieldVector(c = reader.cache["c"], f = reader.cache["f"])
```
Comment (simonbyrne) [inline src/IO/readers.jl:182]: we don't want to hardcode names

**Comment by sriharshakandala** [comment]
bors try

**Comment by sriharshakandala** [comment]
> Does this work for with MPI runs?

The current PR is intended for use with serial runs.
We can temporarily try the case where each process writes its data to a separate file, while parallel HDF5 support is being added.  However, parallel HDF5 is recommended for parallel runs!

**Review by jiahe23 — APPROVED** [review]
looks good. please clean up commit history and merge

**Comment by LenkaNovak** [comment]
I'm trying to run this from `ClimaAtmos` and getting an error because `IO` the module name clashes with `IO` the DataType. I then tried `restart_Y = ClimaCore.IO.read("testfile.hdf5", "Y")` and this doesn't recognize ClimaCore with
Screen Shot 2022-07-13 at 5 59 49 PM
:

**Comment by LenkaNovak** [comment]
It's also strange that it's throwing the error [here](https://github.com/CliMA/ClimaCore.jl/blob/dcbedcc09a98487ba0ddf6642fd31acb27cdec59/src/IO/readers.jl#L146). I'm happy to connect tomorrow, but it would be good to see that Atmos state can be saved / read before merging. (unless Jia's already checked and I'm doing something I shouldn't ;) ). Otherwise the code looks very neat and legible.

**Review by LenkaNovak — CHANGES_REQUESTED** [review]
(see comments above)

**Comment by sriharshakandala** [comment]
I am open to a different name.  We do plan to overload `read/write` to support other formats as well (e.g.: custom binary) in the future. cc: @simonbyrne @bischtob

File: src/IO/IO.jl [hunk 6]
```diff
@@ -0,0 +1,6 @@
+module IO
```
Comment (simonbyrne) [inline src/IO/IO.jl:1]: Following discussion in the standup meeting, lets rename this to be `InputOutput`

File: src/IO/writers.jl [hunk 7]
```diff
@@ -0,0 +1,391 @@
...
+    domain::Domains.IntervalDomain,
+    name::AbstractString = defaultname(domain),
+)
+    group = create_group(writer.file, name)
+    write_attribute(group, "type", "IntervalDomain")
+    write_attribute(group, "coord_type", string(typeof(domain.coord_min)))
+    write_attribute(group, "coord_min", string(domain.coord_min))
+    write_attribute(group, "coord_max", string(domain.coord_max))
+    !isnothing(domain.boundary_names) &&
+        write_attribute(group, "boundary_names", string(domain.boundary_names))
```
Comment (simonbyrne) [inline src/IO/writers.jl:73]: write this as a collection of strings, not a single string
```suggestion
        write_attribute(group, "boundary_names", map(string, domain.boundary_names)))
```

File: src/IO/readers.jl [hunk 8]
```diff
@@ -0,0 +1,356 @@
...
+function read_new!(
+    reader::HDF5Reader,
+    name::AbstractString,
+    ::Type{IntervalDomain},
+)
+    g = reader.file["domains/" * name]
+    coord_type = replace(attrs(g)["coord_type"], "ClimaCore." => "")
+    boundary_names =
+        haskey(attributes(g), "boundary_names") ?
+        eval(Meta.parse(attrs(g)["boundary_names"])) : nothing
```
Comment (simonbyrne) [inline src/IO/readers.jl:105]: by writing this as a collection of strings, you should be able to do this without `eval`

File: src/IO/readers.jl [hunk 9]
```diff
@@ -0,0 +1,356 @@
...
+    periodic = boundary_names === nothing ? true : false
+    CoordType = eval(Meta.parse(coord_type))
```
Comment (simonbyrne) [inline src/IO/readers.jl:107]: There are only handful of coord types: can we just make this an if-else branch instead?

File: src/IO/writers.jl [hunk 10]
```diff
@@ -0,0 +1,391 @@
...
+    writer::HDF5Writer,
+    mesh::Meshes.IntervalMesh,
+    name::AbstractString = defaultname(mesh),
+)
+    meshname = split(name, "/")[end]
+    group = create_group(writer.file, name)
+    write_attribute(group, "type", "IntervalMesh")
+    write!(writer, mesh.domain, "domains/$meshname")
+    write_attribute(group, "nelements", length(mesh.faces))
+    write_attribute(group, "faces", string(mesh.faces))
```
Comment (simonbyrne) [inline src/IO/writers.jl:114]: if `faces` is an `Array`, it should be written as an array. If it's a range, we shouldn't need to write it?

File: src/IO/readers.jl [hunk 11]
```diff
@@ -0,0 +1,356 @@
...
+function read_new!(
+    reader::HDF5Reader,
+    name::AbstractString,
+    ::Type{IntervalMesh},
+)
+    (; file, cache) = reader
+    g = file["meshes/" * name]
+    FT = attrs(g)["FT"] # Float type
+    faces_string = replace(attrs(g)["faces"], "ZPoint" => "ZPoint{$FT}")
+    faces = eval(Meta.parse(faces_string))
```
Comment (simonbyrne) [inline src/IO/readers.jl:146]: Again, if we write it as an `Array` we shouldn't need to `eval` here

File: src/IO/readers.jl [hunk 12]
```diff
@@ -0,0 +1,356 @@
...
+Reads an object named 'name' of type 'EquiangularCubedSphere'.
+"""
+function read_new!(
+    reader::HDF5Reader,
+    name::AbstractString,
+    ::Type{EquiangularCubedSphere},
+)
+    (; file, cache) = reader
+    g = file["meshes/" * name]
+    localelementmap = eval(Meta.parse(attrs(g)["localelementmap"]))
```
Comment (simonbyrne) [inline src/IO/readers.jl:186]: use a branch instead of `eval`

File: src/IO/readers.jl [hunk 13]
```diff
@@ -0,0 +1,356 @@
...
+    ::Type{SpectralElementSpace1D},
+)
+    (; file, cache) = reader
+    g = file["spaces/" * name]
+    quadrature_num_points = attrs(g)["quadrature_num_points"]
+    quadrature_style = eval(
+        Meta.parse(
+            attrs(g)["quadrature_type"] * "{$(quadrature_num_points)}()",
+        ),
+    )
```
Comment (simonbyrne) [inline src/IO/readers.jl:248]: again, branch instead of `eval`

File: src/IO/readers.jl [hunk 14]
```diff
@@ -0,0 +1,356 @@
...
+    ::Type{SpectralElementSpace2D},
+)
+    (; file, cache) = reader
+    g = file["spaces/" * name]
+    quadrature_num_points = attrs(g)["quadrature_num_points"]
+    quadrature_style = eval(
+        Meta.parse(
+            attrs(g)["quadrature_type"] * "{$(quadrature_num_points)}()",
+        ),
+    )
```
Comment (simonbyrne) [inline src/IO/readers.jl:274]: branch instead of `eval` (can write a helper function to avoid duplicating code with above)

File: src/IO/readers.jl [hunk 15]
```diff
@@ -0,0 +1,356 @@
...
+Reads an object named 'name' of type 'ExtrudedFiniteDifferenceSpace'.
+"""
+function read_new!(
+    reader::HDF5Reader,
+    name::AbstractString,
+    ::Type{T},
+) where {T <: ExtrudedFiniteDifferenceSpace}
+    (; file, cache) = reader
+    g = file["spaces/" * name]
+    staggering = eval(Meta.parse(attrs(g)["staggering"]))
```
Comment (simonbyrne) [inline src/IO/readers.jl:295]: if-else branch insetad of eval

File: src/IO/readers.jl [hunk 16]
```diff
@@ -0,0 +1,356 @@
...
+function read_new!(reader::HDF5Reader, name::AbstractString, ::Type{Field})
+    (; file, cache) = reader
+    @assert name ∈ keys(file["fields"])
+    g = file["fields/" * name]
+    data = read_dataset(g, "data")
+    data_layout = attrs(g)["data_layout"]
+    Nij = size(data, findfirst("I", data_layout)[1])
+    DataLayout = eval(Meta.parse(data_layout))
+    ElType =
+        eval(Meta.parse(replace(attrs(g)["value_type"], "ClimaCore." => "")))
```
Comment (simonbyrne) [inline src/IO/readers.jl:321]: can we use an `if-else` branch instead of `eval`?

File: src/IO/writers.jl (hunk 10 above)
Comment (sriharshakandala) [reply src/IO/writers.jl]: `faces` can be range. We we like it to be written as an `Array` anyways?

File: src/IO/readers.jl (hunk 16 above)
Comment (sriharshakandala) [reply src/IO/readers.jl]: `eval` statement has been replaced for `DataLayout`. `ElType` is much more complex to predict!

File: src/IO/writers.jl (hunk 10 above)
Comment (sriharshakandala) [reply src/IO/writers.jl]: Done. `Range` is used to construct a uniform mesh.

File: src/IO/readers.jl (hunk 11 above)
Comment (sriharshakandala) [reply src/IO/readers.jl]: A test is added for stretched mesh. However, the last element of the stretched mesh seems to be a NaN, causing problems with the test.

Comment (sriharshakandala) [reply src/IO/readers.jl]: Bug fix added, resolved.

File: src/Meshes/interval.jl [hunk 17]
```diff
@@ -123,7 +123,7 @@ function IntervalMesh(
     cmax = Geometry.component(domain.coord_max, 1)
     R = cmax - cmin
     h = stretch.H / R
-    η(ζ) = -h * log1p((expm1(-1 / h)) * ζ)
+    η(ζ) = ζ == 1 ? ζ : -h * log1p((expm1(-1 / h)) * ζ)
```
Comment (charleskawczynski) [inline src/Meshes/interval.jl:126]: Can these changes, if intentional, be moved to a separate PR?

Comment (sriharshakandala) [reply src/Meshes/interval.jl:126]: This changes were add by @simonbyrne , to fix a `NaN` issue in the ExponentialStretchedGrid code!

File: test/InputOutput/hybrid2dbox_stretched.jl [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+function init_dry_rising_bubble_2d(x, z)
+    x_c = 0.0
+    z_c = 350.0
+    r_c = 250.0
+    θ_b = 300.0
+    θ_c = 0.5
+    cp_d = C_p
+    cv_d = C_v
+    p_0 = MSLP
+    g = grav
```
[the remark this answers is not in the document]
Comment (sriharshakandala) [reply test/InputOutput/hybrid2dbox_stretched.jl:66]: Removed global variables.

**Comment by simonbyrne** [comment]
Maybe we need a mechanism to close the writers and readers?

**Comment by simonbyrne** [comment]
e.g. define `Base.close` for each, which will close the underlying HDF5 handles

**Comment by sriharshakandala** [comment]
bors try

**Comment by sriharshakandala** [comment]
bors try

**Comment by sriharshakandala** [comment]
bors try

**Comment by sriharshakandala** [comment]
bors try

File: .buildkite/pipeline.yml [hunk 19]
```diff
@@ -25,6 +25,7 @@ steps:
       - "julia --project=perf -e 'using Pkg; Pkg.precompile(;strict=true)'"
 
       - echo "--- Instantiate examples"
+      - "julia --project=perf -e 'using Pkg; Pkg.develop(path=\".\")'"
```
Comment (charleskawczynski) [inline .buildkite/pipeline.yml:28]: Please remove this, our perf environment has the manifest checked in.

**Comment by sriharshakandala** [comment]
bors try

**Comment by charleskawczynski** [comment]
I think we just need to

```
julia --project=docs/
] add HDF5
```
?

**Comment by charleskawczynski** [comment]
Also, can we rebase/squash?

**Review by charleskawczynski — APPROVED** [review]

File: docs/Manifest.toml [hunk 20]
```diff
@@ -200,7 +200,7 @@ uuid = "3a4d1b5c-c61d-41fd-a00a-5873ba7a1b0d"
 version = "0.3.0"
 
 [[deps.ClimaCore]]
-deps = ["Adapt", "BlockArrays", "CUDA", "ClimaComms", "CubedSphere", "DataStructures", "DiffEqBase", "DocStringExtensions", "ForwardDiff", "GaussQuadrature", "GilbertCurves", "InteractiveUtils", "IntervalSets", "LinearAlgebra", "Plots", "RecursiveArrayTools", "RootSolvers", "Rotations", "SparseArrays", "Static", "StaticArrays", "Statistics", "UnPack"]
+deps = ["Adapt", "BlockArrays", "CUDA", "ClimaComms", "CubedSphere", "DataStructures", "DiffEqBase", "DocStringExtensions", "ForwardDiff", "GaussQuadrature", "GilbertCurves", "HDF5", "InteractiveUtils", "IntervalSets", "LinearAlgebra", "RecursiveArrayTools", "RootSolvers", "Rotations", "SparseArrays", "Static", "StaticArrays", "Statistics", "UnPack"]
```
Comment (charleskawczynski) [inline docs/Manifest.toml:203]: It looks like Plots was removed here?

File: examples/Manifest.toml [hunk 21]
```diff
@@ -160,7 +160,7 @@ uuid = "5f86816e-8b66-43b2-912e-75384f99de49"
 version = "0.3.2"
 
 [[deps.ClimaCore]]
-deps = ["Adapt", "BlockArrays", "CUDA", "ClimaComms", "CubedSphere", "DataStructures", "DiffEqBase", "DocStringExtensions", "ForwardDiff", "GaussQuadrature", "GilbertCurves", "InteractiveUtils", "IntervalSets", "LinearAlgebra", "Plots", "RecursiveArrayTools", "RootSolvers", "Rotations", "SparseArrays", "Static", "StaticArrays", "Statistics", "UnPack"]
+deps = ["Adapt", "BlockArrays", "CUDA", "ClimaComms", "CubedSphere", "DataStructures", "DiffEqBase", "DocStringExtensions", "ForwardDiff", "GaussQuadrature", "GilbertCurves", "HDF5", "InteractiveUtils", "IntervalSets", "LinearAlgebra", "RecursiveArrayTools", "RootSolvers", "Rotations", "SparseArrays", "Static", "StaticArrays", "Statistics", "UnPack"]
```
Comment (charleskawczynski) [inline examples/Manifest.toml:163]: It looks like Plots was removed here?

Comment (charleskawczynski) [reply examples/Manifest.toml:163]: Maybe this is the issue?

**Comment by charleskawczynski** [comment]
It looks like there's nothing wrong in the main branch. Something wrong is happening with this PR. Looking at the log, this seems suspect:

```julia
┌ Warning: Replacing module ClimaCore
```

in the `Column Heat Diffusion Eq` job. It looks like there's still use of `eval`, which I would recommend looking into first. I'm still looking into this, but just wanted to point this out in case it's helpful to others in finding out what's wrong.

I also recommend insuring that `Plots` remains part of the `examples` and `docs` Manifest, this PR should probably just add `HDF5`

File: src/InputOutput/readers.jl [hunk 22]
```diff
@@ -0,0 +1,419 @@
...
+function read_new!(reader::HDF5Reader, name::AbstractString, ::Type{Field})
+    (; file, cache) = reader
+    @assert name ∈ keys(file["fields"])
+    g = file["fields/" * name]
+    data = read_dataset(g, "data")
+    data_layout = attrs(g)["data_layout"]
+    Nij = size(data, findfirst("I", data_layout)[1])
+    DataLayout = _scan_data_layout(data_layout)
+    ElType =
+        eval(Meta.parse(replace(attrs(g)["value_type"], "ClimaCore." => "")))
```
Comment (charleskawczynski) [inline src/InputOutput/readers.jl:378]: Can we avoid `eval` here?

File: test/InputOutput/hybrid3dbox.jl [hunk 23]
```diff
@@ -0,0 +1,124 @@
+using Test
+using StaticArrays, IntervalSets, LinearAlgebra, UnPack
+
+import ClimaCore:
+    ClimaCore,
```
Comment (charleskawczynski) [inline test/InputOutput/hybrid3dbox.jl:5]:
```suggestion
import ClimaCore
import ClimaCore:
```
I don't think that this should matter, but I've not really used this syntax before

Comment (charleskawczynski) [reply test/InputOutput/hybrid3dbox.jl:5]: Also, there are a couple of these

**Comment by sriharshakandala** [comment]
bors try

**Comment by sriharshakandala** [comment]
bors try
